pkg/cmd/pulumi/templates/org.go

```diff
@@ -0,0 +1,221 @@
+				if templateName != "" && templateName != template.Name {
+					logging.V(10).Infof("skipping template %q", template.Name)
+					continue
+				}
+
+				logging.V(10).Infof("adding template %q", template.Name)
+				s.addTemplate(Template{
+					name:               template.Name,
+					description:        "", // No description present.
+					projectDescription: template.Description,
+					source:             s,
+					download: func(ctx context.Context) (workspace.Template, error) {
+						templateDir, err := os.MkdirTemp("", "pulumi-template-")
+						if err != nil {
+							return workspace.Template{}, err
+						}
+						// Having created a template directory, we now add it to the list of directories to close.
+						s.addCloser(func() error { return os.RemoveAll(templateDir) })
+
+						tarReader, err := b.DownloadTemplate(ctx, org, template.TemplateURL)
+						if err != nil {
+							return workspace.Template{}, err
+						}
+						if err := errors.Join(
+							writeTar(ctx, tarReader.Tar(), templateDir),
+							tarReader.Close(),
+						); err != nil {
+							return workspace.Template{}, err
+						}
+						logging.Infof("downloaded %q into %q", template.Name, templateDir)
+
+						return workspace.LoadTemplate(templateDir)
+					},
+				})
+			}
+		}
+	}
+
+	for _, org := range orgs {
+		wg.Add(1)
+		go handleOrg(org)
+	}
+}
+
+func writeTar(_ context.Context, reader *tar.Reader, dst string) error {
+	for {
+		header, err := reader.Next()
+		if errors.Is(err, io.EOF) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		logging.V(8).Infof("Decompressing %q", header.Name)
+
+		path := filepath.Clean(header.Name)
+		if !filepath.IsLocal(path) {
+			return fmt.Errorf("refusing to write non-local path %q", path)
+		}
+
+		target := filepath.Join(dst, path)
+
+		// Ensure that we can write the directory
+		if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
+			return err
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if header.Mode > math.MaxUint32 {
+				return fmt.Errorf("invalid file mode for %q: %02x", header.Name, header.Mode)
+			}
+
+			fileMode := os.FileMode(header.Mode) //nolint:gosec // We checked the overflow
+			err := os.Mkdir(target, fileMode)
+			if err != nil && !errors.Is(err, fs.ErrExist) {
+				return err
+			}
+
+		case tar.TypeReg:
+			if header.Mode > math.MaxUint32 {
+				return fmt.Errorf("invalid file mode for %q: %02x", header.Name, header.Mode)
+			}
+
+			fileMode := os.FileMode(header.Mode) //nolint:gosec // We checked the overflow
+			f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, fileMode)
+			if err != nil {
+				return err
+			}
+
+			// Write the tar file into f
+			if _, err := io.Copy(f, reader); err != nil {
+				return errors.Join(err, f.Close())
+			}
+
+			// Close f. We don't use defer because we want to close each file
+			// after we open it, not leave it all until the download function
+			// exits.
+			if err := f.Close(); err != nil {
```

Comment:
This means we don't close it if `io.Copy` fails. You can instead use a function to scope the defer:

```
	func() {
			f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, fileMode)
			if err != nil {
				return err
			}
			defer f.Close()

			// Write the tar file into f
			if _, err := io.Copy(f, reader); err != nil {
				return errors.Join(err, f.Close())
			}
	}()
```

Reply:
We do close if `io.Copy` fails:

https://github.com/pulumi/pulumi/blob/13700cd0f955f1e51a10edae44ef107de0c13661/pkg/cmd/pulumi/templates/org.go#L209-L211

I changed the implementation to include an immediately invoked closure so it's more obvious that `f.Close` is always called.